src/osm/GeometryHandler.cpp

```diff
@@ -61,137 +79,402 @@ osm2rdf::osm::GeometryHandler<W>::GeometryHandler(
+    envelopes.push_back(area.envelope());
   }
+
+  std::unordered_map<int32_t, osm2rdf::geometry::Area> cutouts;
+
+  const auto& boxIds =
+      pack(getBoxIds(area.geom(), envelopes, innerGeom, outerGeom,
+                     totPoints > MIN_CUTOUT_POINTS ? &cutouts : 0));
+
 #pragma omp critical(areaDataInsert)
   {
     if (area.hasName()) {
-      if (_config.minimalAreaEnvelopeRatio <= 0.0 ||
-          area.geomArea() / area.envelopeArea() >=
-              _config.minimalAreaEnvelopeRatio) {
-        _spatialStorageArea.push_back(
-            SpatialAreaValue(area.envelope(), area.id(), geom, area.objId(),
-                             area.geomArea(), area.fromWay()));
-      } else {
-        // we have bad area envelope proportions -> treat as unnamed area
-        _oaUnnamedAreas << SpatialAreaValue(area.envelope(), area.id(), geom,
-                                            area.objId(), area.geomArea(),
-                                            area.fromWay());
-        _numUnnamedAreas++;
-      }
+      _spatialStorageArea.push_back(
+          {envelopes, area.id(), geom, area.objId(), area.geomArea(),
+           area.fromWay() ? AreaFromType::FROM_WAY : AreaFromType::FROM_REL,
+           innerGeom, outerGeom, boxIds, cutouts, convexHull});
     } else if (!area.fromWay()) {
       // Areas from ways are handled in GeometryHandler<W>::way
-      _oaUnnamedAreas << SpatialAreaValue(area.envelope(), area.id(), geom,
-                                          area.objId(), area.geomArea(),
-                                          area.fromWay());
+      _oaUnnamedAreas << SpatialAreaValue(
+          envelopes, area.id(), geom, area.objId(), area.geomArea(),
+          area.fromWay() ? AreaFromType::FROM_WAY : AreaFromType::FROM_REL,
+          innerGeom, outerGeom, boxIds, cutouts, convexHull);
       _numUnnamedAreas++;
     }
   }
 }
 
 // ____________________________________________________________________________
 template <typename W>
-void osm2rdf::osm::GeometryHandler<W>::node(const osm2rdf::osm::Node& node) {
+void GeometryHandler<W>::node(const Node& node) {
 #pragma omp critical(nodeDataInsert)
   {
-    _oaNodes << SpatialNodeValue(node.envelope(), node.id(), node.geom());
+    _oaNodes << SpatialNodeValue(node.id(), node.geom());
     _numNodes++;
   }
 }
 
 // ____________________________________________________________________________
 template <typename W>
-void osm2rdf::osm::GeometryHandler<W>::way(const osm2rdf::osm::Way& way) {
+void GeometryHandler<W>::way(const Way& way) {
   WayNodeList nodeIds;
   nodeIds.reserve(way.nodes().size());
+
   for (const auto& nodeRef : way.nodes()) {
     nodeIds.push_back(nodeRef.id());
   }
-  auto geom = way.geom();
-  if (_config.simplifyGeometries > 0) {
-    geom = simplifyGeometry(geom);
+  const auto& geom = simplifyGeometry(way.geom());
+
+  std::vector<osm2rdf::geometry::Box> boxes;
+
+  const size_t CHUNKSIZE = 25;
+
+  if (geom.size() >= 2 * CHUNKSIZE) {
```

Comment:
Just to be sure: if a way has at least `2 * CHUNKSIZE` nodes it receives an additional bounding boxes for every parts of size `CHUNKSIZE` (except the last) which is later added to the r-tree? A small comment to answer why this is done would improve readability here by a huge point - this is related to the remark about `GeometryHandler<W>::area` where a short explanation why could improve the understanding too

Reply:
No, it is not added to the R-Tree (ways are never added to the R-Tree). These envelopes are used for *querying* the R-Tree in `indexQryIntersect()`. That is, instead of sending a huge intersect query for the entire bounding box of the way, we send several smaller queries to the R-Tree, aggregate the result and make it unique later on.

Reply:
Makes sense, I quess it was a bit late when I looked at it. But I think adding a small explanation to the code would still be good.

Reply:
I agree, I have already added the comment, will commit in a few minutes :D